engines/wintermute/base/gfx/opengl/x/meshx.cpp

```diff
@@ -0,0 +1,486 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+/*
+ * This file is based on WME.
+ * http://dead-code.org/redir.php?target=wme
+ * Copyright (c) 2003-2013 Jan Nedoma and contributors
+ */
+
+#include "common/math.h"
+#include "engines/wintermute/base/base_game.h"
+#include "engines/wintermute/base/gfx/opengl/base_surface_opengl3d.h"
+#include "engines/wintermute/base/gfx/opengl/material.h"
+#include "engines/wintermute/base/gfx/opengl/shadow_volume.h"
+#include "engines/wintermute/base/gfx/opengl/x/frame_node.h"
+#include "engines/wintermute/base/gfx/opengl/x/meshx.h"
+#include "engines/wintermute/base/gfx/opengl/x/modelx.h"
+#include "engines/wintermute/base/gfx/opengl/x/loader_x.h"
+#include "engines/wintermute/dcgf.h"
+
+namespace Wintermute {
+
+//////////////////////////////////////////////////////////////////////////
+MeshX::MeshX(BaseGame *inGame) : BaseNamedObject(inGame) {
+	_numAttrs = 0;
+	_maxFaceInfluence = 0;
+
+	_adjacency = nullptr;
+	_skinAdjacency = nullptr;
+
+	_BBoxStart = _BBoxEnd = Math::Vector3d(0.0f, 0.0f, 0.0f);
+	_skinnedMesh = true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+MeshX::~MeshX() {
+	delete[] _adjacency;
+	delete[] _skinAdjacency;
+	delete[] _vertexPositionData;
+
+	for (uint32 i = 0; i < _materials.size(); i++) {
+		delete _materials[i];
+	}
+
+	_materials.clear();
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::loadFromX(const Common::String &filename, XFileLexer &lexer) {
+	bool res = true;
+
+	lexer.advanceOnOpenBraces();
+	_vertexCount = lexer.tokenToInt();
+	lexer.next();
+	lexer.next();
+
+	// vertex format for .X meshes will be position + normals + textures
+	const int vertexSize = 32;
+	_vertexData = new byte[vertexSize * _vertexCount];
+	_vertexPositionData = new float[3 * _vertexCount];
+	float *vertexCoords = reinterpret_cast<float *>(_vertexData);
+
+	for (uint32 i = 0; i < _vertexCount; ++i) {
+		for (int j = 0; j < 3; ++j) {
+			_vertexPositionData[i * 3 + j] = lexer.tokenToFloat();
+			vertexCoords[i * 8 + 5 + j] = _vertexPositionData[i * 3 + j];
+			lexer.next();
+			lexer.next();
+		}
+		lexer.next();
+	}
+
+	int faceCount = lexer.tokenToInt();
+	lexer.next();
+	lexer.next();
+	// we should be able to assume for now that
+	// we are only dealing with triangles
+	_indexData = new uint16[faceCount * 3];
+	_indexCount = faceCount * 3;
+
+	for (int i = 0; i < faceCount; ++i) {
+		int indexCount = lexer.tokenToInt();
+		lexer.next();
+		lexer.next();
+
+		if (indexCount != 3) {
+			warning("MeshX::loadFromX non triangle faces are not supported");
+			return false;
+		}
+
+		for (int j = 0; j < 3; ++j) {
+			_indexData[i * 3 + j] = lexer.tokenToInt();
+			lexer.next();
+			lexer.next();
+		}
+
+		lexer.next();
+	}
+
+	while (!lexer.eof()) {
+		if (lexer.tokenIsIdentifier("MeshTextureCoords")) {

```

Comment:
I would suggest putting each of these cases into a separate function, to make this function a bit more manageable.

Reply:
Will definitively do

Reply:
Done